pkg/apis/application/v1alpha1/types.go

```diff
@@ -0,0 +1,1279 @@
+	// This should only be changed in exceptional circumstances.
+	// Setting to zero will store no history. This will reduce storage used.
+	// Increasing will increase the space used to store the history, so we do not recommend increasing it.
+	// Default is 10.
+	RevisionHistoryLimit *int64 `json:"revisionHistoryLimit,omitempty" protobuf:"bytes,7,name=revisionHistoryLimit"`
+}
+
+// ResourceIgnoreDifferences contains resource filter and list of json paths which should be ignored during comparison with live state.
+type ResourceIgnoreDifferences struct {
+	Group             string   `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+	Kind              string   `json:"kind" protobuf:"bytes,2,opt,name=kind"`
+	Name              string   `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+	Namespace         string   `json:"namespace,omitempty" protobuf:"bytes,4,opt,name=namespace"`
+	JSONPointers      []string `json:"jsonPointers,omitempty" protobuf:"bytes,5,opt,name=jsonPointers"`
+	JQPathExpressions []string `json:"jqPathExpressions,omitempty" protobuf:"bytes,6,opt,name=jqPathExpressions"`
+}
+
+// EnvEntry represents an entry in the application's environment
+type EnvEntry struct {
+	// Name is the name of the variable, usually expressed in uppercase
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Value is the value of the variable
+	Value string `json:"value" protobuf:"bytes,2,opt,name=value"`
+}
+
+// IsZero returns true if a variable is considered empty or unset
+func (a *EnvEntry) IsZero() bool {
+	return a == nil || a.Name == "" && a.Value == ""
+}
+
+// NewEnvEntry parses a string in format name=value and returns an EnvEntry object
+func NewEnvEntry(text string) (*EnvEntry, error) {
+	parts := strings.SplitN(text, "=", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("Expected env entry of the form: param=value. Received: %s", text)
+	}
+	return &EnvEntry{
+		Name:  parts[0],
+		Value: parts[1],
+	}, nil
+}
+
+// Env is a list of environment variable entries
+type Env []*EnvEntry
+
+// IsZero returns true if a list of variables is considered empty
+func (e Env) IsZero() bool {
+	return len(e) == 0
+}
+
+// Environ returns a list of environment variables in name=value format from a list of variables
+func (e Env) Environ() []string {
+	var environ []string
+	for _, item := range e {
+		if !item.IsZero() {
+			environ = append(environ, fmt.Sprintf("%s=%s", item.Name, item.Value))
+		}
+	}
+	return environ
+}
+
+// Envsubst interpolates variable references in a string from a list of variables
+func (e Env) Envsubst(s string) string {
+	valByEnv := map[string]string{}
+	for _, item := range e {
+		valByEnv[item.Name] = item.Value
+	}
+	return os.Expand(s, func(s string) string {
+		return valByEnv[s]
+	})
+}
+
+// ApplicationSource contains all required information about the source of an application
+type ApplicationSource struct {
```

Comment:
In general it's a bit more of copy/paste that I hoped it to be. Are you sure it will be ok to maintain?

I thought we have only the essential types copied (like this one here) without any methods and only couple of consts and we are good to go.

But if that's easier that way then I'm fine with that.

Reply:
I can trim it down more. I can trim it down more (methods, unused fields).

Reply:
Questions are: Does it make sense? Will it make maintenance easier?

Reply:
I'd rather support ~180 lines than ~1200 lines, so good point.